### PR TITLE
[SPARK-32951][SQL] Foldable propagation from Aggregate

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?
This PR adds foldable propagation from `Aggregate` as per: https://github.com/apache/spark/pull/29771#discussion_r490412031

### Why are the changes needed?
This is an improvement as `Aggregate`'s `aggregateExpressions` can contain foldables that can be propagated up.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
New UT.

### CODE REVIEW

**Comment by peter-toth** [comment]
cc @cloud-fan, @dongjoon-hyun, @gatorsmile, @maropu, @rednaxelafx, @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:724]: can we merge this comment with ```// Handling `Project` is moved to `propagateFoldables`.```?

**Review by cloud-fan — APPROVED** [review]
LGTM if tests can pass

**Comment by peter-toth** [comment]
> LGTM if tests can pass

I need to look into plan stability failures first.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:724]: Ok, will change it.

**Comment by cloud-fan** [comment]
you can just regenerate the golden file and push the change. Then the reviewers can look at the plan change and approve it if it's fine.

**Comment by peter-toth** [comment]
> you can just regenerate the golden file and push the change. Then the reviewers can look at the plan change and approve it if it's fine.

Thanks. I did it already but haven't had time to check the changes. Anyway, added it as a new commit.

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q41.sf100/explain.txt
Comment (cloud-fan) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q41.sf100/explain.txt:76]: This file is a positive change.

**Review by viirya — APPROVED** [review]

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q41.sf100/explain.txt
Comment (viirya) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q41.sf100/explain.txt:84]: this looks good.

**Comment by viirya** [comment]
Please remove [WIP] before merging.

**Comment by peter-toth** [comment]
> Please remove [WIP] before merging.

Done, thanks.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @peter-toth , @cloud-fan , and @viirya .
Merged to master for Apache Spark 3.1.0 on December 2020.

**Comment by maropu** [comment]
late LGTM and thanks for the update, @peter-toth !

**Pull request closed without merging**
